### PR TITLE
llama-cpp: init at unstable-2023-04-06

### PR DESCRIPTION
**Pull request by dit7ya**
###### Description of changes

Makes progress towards #224921.

https://github.com/ggerganov/llama.cpp/

NOTE: I have not tested all of the scripts and binaries. Please test and report if something is off.

###### Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

**Comment by SomeoneSerge** [comment]
Also cf https://github.com/SomeoneSerge/pkgs/tree/master/pkgs/llama.cpp and we really should get cmake install targets merged upstream

**Comment by MostAwesomeDude** [comment]
Works for me. Here's a possible way to reproduce:

* Grab `llama7b-4bit.pt` files from [here](https://huggingface.co/maderix/llama-65b-4bit/tree/main)
* Grab `tokenizer.model` from [here](https://huggingface.co/nyanko7/LLaMA-7B/tree/main)
* `result/bin/convert-gptq-to-ggml llama7b-4bit.pt tokenizer.model 7b-4bit.bin`
* Apply [this fix](https://github.com/ggerganov/llama.cpp/pull/770/files) by hand
* `result/bin/llama-cpp -m 7b-4bit.bin -p "With just 10 simple steps, anybody can build a website:" -n 512`

It would be nice to grab a commit which doesn't need the magic numbers to be patched, but there might not be a better candidate.

File: pkgs/tools/misc/llama-cpp/default.nix [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+    for f in $out/libexec/*.py; do
+      makeWrapper ${pythonEnv}/bin/python "$out/bin/$(basename $f .py)" \
+        --add-flags "$f"
+    done
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "Port of Facebook's LLaMA model in C/C";
```
Comment (nagy) [inline pkgs/tools/misc/llama-cpp/default.nix:55]: "in C/C"  -> "in C/C++" ?

**Review by MostAwesomeDude — APPROVED** [review]

**Comment by dit7ya** [comment]
Actually, let's wait for https://github.com/ggerganov/llama.cpp/pull/545 to be merged.

**Comment by MostAwesomeDude** [comment]
@prusnak I have integrated the upstream flake into my project. I have already found a hitch: I need `libllama.so`, which isn't built by that flake. I'm going to submit a fix upstream, but I have no way to test whether Darwin support, particularly `aarch64-darwin` support, will be affected. I mention all of this to highlight how the nixpkgs process compares to flakes.

File: pkgs/tools/misc/llama-cpp/default.nix [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+    cp ./bin/perplexity $out/bin/llama-cpp-perplexity
+    cp ./bin/embedding $out/bin/llama-cpp-embedding
+    cp ./bin/quantize $out/bin/llama-cpp-quantize
+
+    cp ../*.py $out/libexec/
+
+    for f in $out/libexec/*.py; do
+      makeWrapper ${pythonEnv}/bin/python "$out/bin/$(basename $f .py)" \
+        --add-flags "$f"
+    done
```
Comment (SuperSandro2000) [inline pkgs/tools/misc/llama-cpp/default.nix:49]: We can't use one of the python wrappers here?

**Comment by Freed-Wu** [comment]
It's out of date. I try to update flake.nix to make it newer. <https://github.com/ggerganov/llama.cpp/pull/2270>

```
❯ tree result/
 result
├──  bin
│  ├──  baby-llama
│  ├──  benchmark
│  ├──  convert.py
│  ├──  embd-input-test
│  ├──  embedding
│  ├──  llama
│  ├──  llama-server
│  ├──  perplexity
│  ├──  q8dot
│  ├──  quantize
│  ├──  quantize-stats
│  ├──  save-load-state
│  ├──  simple
│  ├──  test-grad0
│  ├──  test-quantize-fns
│  ├──  test-quantize-perf
│  ├──  test-sampling
│  ├──  test-tokenizer-0
│  ├──  train-text-from-scratch
│  └──  vdot
└──  lib
   ├──  libggml_shared.so
   └──  libllama.so
```

> since it is very rapidly evolving.

Right, very very rapid.

**Comment by prusnak** [comment]
> It's out of date. I try to update flake.nix to make it newer. [ggerganov/llama.cpp#2270](https://github.com/ggerganov/llama.cpp/pull/2270)

Since the upstream is very responsive in accepting PRs to `flake.nix` and since the development of upstream is very rapid, I propose to close this PR in favor of using upstream `flake.nix`

**Comment by hughobrien** [comment]
> Since the upstream is very responsive in accepting PRs to `flake.nix` and since the development of upstream is very rapid, I propose to close this PR in favor of using upstream `flake.nix`

Seconded, the pace of development is incredible.

**Pull request closed without merging**